pvlib/modelchain.py

```diff
@@ -1407,13 +1733,66 @@ def run_model_from_effective_irradiance(self, data=None):
 
         See also
         --------
-        pvlib.modelchain.ModelChain.run_model_from
+        pvlib.modelchain.ModelChain.run_model
         pvlib.modelchain.ModelChain.run_model_from_poa
         """
-
+        self._check_multiple_input(data)
         self._assign_weather(data)
         self._assign_total_irrad(data)
-        self.effective_irradiance = data['effective_irradiance']
+        self.results.effective_irradiance = _tuple_from_dfs(
+            data, 'effective_irradiance')
         self._run_from_effective_irrad(data)
 
         return self
+
+
+def _snl_params(inverter_params):
+    """Return True if `inverter_params` includes parameters for the
+    Sandia inverter model."""
+    return {'C0', 'C1', 'C2'} <= inverter_params
+
+
+def _adr_params(inverter_params):
+    """Return True if `inverter_params` includes parameters for the ADR
+    inverter model."""
+    return {'ADRCoefficients'} <= inverter_params
+
+
+def _pvwatts_params(inverter_params):
+    """Return True if `inverter_params` includes parameters for the
+    PVWatts inverter model."""
+    return {'pdc0'} <= inverter_params
+
+
+def _copy(data):
+    """Return a copy of each DataFrame in `data` if it is a tuple,
+    otherwise return a copy of `data`."""
+    if not isinstance(data, tuple):
+        return data.copy()
+    return tuple(df.copy() for df in data)
+
+
+def _all_same_index(data):
+    indexes = map(lambda df: df.index, data)
+    next(indexes, None)
+    for index in indexes:
+        if not index.equals(data[0].index):
+            raise ValueError("Input DataFrames must have same index.")
+
+
+def _common_keys(dicts):
```

Comment:
This function should work with series and tuples of series, too. There's a documentation failure in the introtutorial.rst due to this, but I think a lot of users would run into it since they're likely to just pull a series from the module/inverter dataframes.

```python
modelchain._common_keys({'a': 1})

{'a'}

modelchain._common_keys(pd.Series({'a': 1}))

{1}

modelchain._common_keys((pd.Series({'a': 1}), pd.Series({'a': 1})))

{1}
```

Alternatively we could enforce the type of `Array.module_parameters` and `Array.inverter_parameters`, but I don't see other advantages to that at this time.

Reply:
Just to be clear, we should allow a tuple of `Series` as input and return the intersection of the indices?

Reply:
> Just to be clear, we should allow a tuple of Series as input and return the intersection of the indices?

Yes.

I'd also expect it work with any "dict-like" objects, so this should be valid:

```python
>>> modelchain._common_keys((pd.Series({'a': 1}), {'a': 1, 'b': 2}))
{'a'}
```

but I'm not sure if this case is in the uncommon or essentially never category.

Reply:
I've written it in a way that supports the example above and included a new test for those types of mixed inputs. I'm not sure about _any_ dict-like type, but as written it will support any type with a `keys()` method. This covers `dict`, `Series`, and `DataFrame`.